# SQL Basics

## Introduction

This notebook is a short introduction to SQL. It is based on the [DuckDB](https://duckdb.org/) database engine.

## Datasets

The following datasets are used in this notebook. You don't need to download them, they can be accessed directly from the notebook.

- [cities.csv](https://open.gishub.org/data/duckdb/cities.csv)
- [countries.csv](https://open.gishub.org/data/duckdb/countries.csv)

## References

- [W3Schools SQL Tutorial](https://www.w3schools.com/sql)
- [DuckDB SQL Introduction](https://duckdb.org/docs/sql/introduction.html)

## Installation

Uncomment the following cell to install the required packages.

In [6]:
%pip install duckdb duckdb-engine jupysql

Note: you may need to restart the kernel to use updated packages.


## Library Import and Configuration

In [7]:
import duckdb
import pandas as pd

# Import jupysql Jupyter extension to create SQL cells
%load_ext sql

Set configurations on jupysql to directly output data to Pandas and to simplify the output that is printed to the notebook.

In [8]:
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

## Connecting to DuckDB

Connect jupysql to DuckDB using a SQLAlchemy-style connection string. You may either connect to an in memory DuckDB, or a file backed db.

In [9]:
#创建一个 DuckDB 内存数据库,一旦 Notebook 关闭，数据就会丢失
%sql duckdb:///:memory:

# %sql duckdb:///path/to/file.db

 If your SQL query is one line only, you may use the `%sql` magic command. For multi-line SQL query, you may use the `%%sql` magic command.

## Install extensions


Check available DuckDB extensions.

In [ ]:
%%sql

SELECT * FROM duckdb_extensions();

extension_name  loaded  installed  \
0              arrow   False      False   
1       autocomplete   False      False   
2                aws   False      False   
3              azure   False      False   
4              excel   False      False   
5                fts   False       True   
6             httpfs   False       True   
7            iceberg   False      False   
8                icu    True       True   
9               inet   False      False   
10          jemalloc   False      False   
11              json    True       True   
12        motherduck   False      False   
13     mysql_scanner   False      False   
14           parquet    True       True   
15  postgres_scanner   False      False   
16           spatial   False       True   
17    sqlite_scanner   False      False   
18         substrait   False      False   
19             tpcds    True       True   
20              tpch    True       True   
21        visualizer   False      False   

                                         install_path  \
0                                                       
1                                                       
2                                                       
3                                                       
4                                                       
5                                          (BUILT-IN)   
6   /Users/ollieballinger/.duckdb/extensions/v0.9....   
7                                                       
8                                          (BUILT-IN)   
9                                                       
10                                                      
11                                         (BUILT-IN)   
12                                                      
13                                                      
14                                         (BUILT-IN)   
15                                                      
16  /Users/ollieballinger/.duckdb/extensions/v0.9....   
17                                                      
18                                                      
19                                         (BUILT-IN)   
20                                         (BUILT-IN)   
21                                                      

                                          description            aliases  
0   A zero-copy data integration between Apache Ar...                 []  
1          Adds support for autocomplete in the shell                 []  
2        Provides features that depend on the AWS SDK                 []  
3   Adds a filesystem abstraction for Azure blob s...                 []  
4          Adds support for Excel-like format strings                 []  
5           Adds support for Full-Text Search Indexes                 []  
6   Adds support for reading and writing files ove...  [http, https, s3]  
7                     Adds support for Apache Iceberg                 []  
8   Adds support for time zones and collations usi...                 []  
9   Adds support for IP-related data types and fun...                 []  
10          Overwrites system allocator with JEMalloc                 []  
11                   Adds support for JSON operations                 []  
12     Enables motherduck integration with the system               [md]  
13    Adds support for connecting to a MySQL database            [mysql]  
14  Adds support for reading and writing parquet f...                 []  
15  Adds support for connecting to a Postgres data...         [postgres]  
16  Geospatial extension that adds support for wor...                 []  
17  Adds support for reading and writing SQLite da...  [sqlite, sqlite3]  
18         Adds support for the Substrait integration                 []  
19      Adds TPC-DS data generation and query support                 []  
20       Adds TPC-H data generation and query support                 []  
21  Creates an HTML-based visualization of the que... 

DuckDB's [httpfs extension](https://duckdb.org/docs/extensions/httpfs) allows parquet and csv files to be queried remotely over http. This is useful for querying large datasets without having to download them locally. Let's install the extension and load the extension.

In [10]:
%%sql

INSTALL httpfs;
LOAD httpfs;

Empty DataFrame
Columns: [Success]
Index: []

## Read CSV

Use the `httpfs` extension to read the `cities.csv` file from the web.

In [11]:
%%sql

SELECT * FROM 'https://open.gishub.org/data/duckdb/cities.csv';

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

In [12]:
%%sql

SELECT * FROM 'https://open.gishub.org/data/duckdb/countries.csv';

id            Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0      1        Afghanistan          AF         AFG             4   33.0000   
1      2            Albania          AL         ALB             8   41.0000   
2      3            Algeria          DZ         DZA            12   28.0000   
3      4     American Samoa          AS         ASM            16  -14.3333   
4      5            Andorra          AD         AND            20   42.5000   
..   ...                ...         ...         ...           ...       ...   
238  239  Wallis and Futuna          WF         WLF           876  -13.3000   
239  240     Western Sahara          EH         ESH           732   24.5000   
240  241              Yemen          YE         YEM           887   15.0000   
241  242             Zambia          ZM         ZMB           894  -15.0000   
242  243           Zimbabwe          ZW         ZWE           716  -20.0000   

     Longitude  
0         65.0  
1         20.0  
2          3.0  
3       -170.0  
4          1.6  
..         ...  
238     -176.2  
239      -13.0  
240       48.0  
241       30.0  
242       30.0  

[243 rows x 7 columns]

## Create Table

Create a table named `cities` from the `cities.csv` file.

In [13]:
%%sql

CREATE TABLE cities AS SELECT * FROM 'https://open.gishub.org/data/duckdb/cities.csv';

Empty DataFrame
Columns: [Success]
Index: []

Create a table named `countries` from the `countries.csv` file.

In [14]:
%%sql

CREATE TABLE countries AS SELECT * FROM 'https://open.gishub.org/data/duckdb/countries.csv';

Empty DataFrame
Columns: [Success]
Index: []

Display the table content in the database.

In [15]:
%%sql

FROM cities;

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

In [16]:
%%sql

FROM countries;

id            Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0      1        Afghanistan          AF         AFG             4   33.0000   
1      2            Albania          AL         ALB             8   41.0000   
2      3            Algeria          DZ         DZA            12   28.0000   
3      4     American Samoa          AS         ASM            16  -14.3333   
4      5            Andorra          AD         AND            20   42.5000   
..   ...                ...         ...         ...           ...       ...   
238  239  Wallis and Futuna          WF         WLF           876  -13.3000   
239  240     Western Sahara          EH         ESH           732   24.5000   
240  241              Yemen          YE         YEM           887   15.0000   
241  242             Zambia          ZM         ZMB           894  -15.0000   
242  243           Zimbabwe          ZW         ZWE           716  -20.0000   

     Longitude  
0         65.0  
1         20.0  
2          3.0  
3       -170.0  
4          1.6  
..         ...  
238     -176.2  
239      -13.0  
240       48.0  
241       30.0  
242       30.0  

[243 rows x 7 columns]

## The SQL SELECT statement

The `SELECT` statement is used to select data from a database. Use either `SELECT *` to select all columns, or `SELECT column1, column2, ...` to select specific columns.

`SELECT * FROM cities` is the same as `FROM cities`.

In [ ]:
%%sql

SELECT * FROM cities;

To limit the number of rows returned, use the `LIMIT` keyword. For example, `SELECT * FROM cities LIMIT 10` will return only the first 10 rows.

In [17]:
%%sql

SELECT * FROM cities LIMIT 10;

id              name country  latitude  longitude  population
0   1             Bombo     UGA   0.58330   32.53330       75000
1   2       Fort Portal     UGA   0.67100   30.27500       42670
2   3           Potenza     ITA  40.64200   15.79900       69060
3   4        Campobasso     ITA  41.56300   14.65600       50762
4   5             Aosta     ITA  45.73700    7.31500       34062
5   6         Mariehamn     ALD  60.09700   19.94900       10682
6   7          Ramallah     PSE  31.90294   35.20621       24599
7   8      Vatican City     VAT  41.90001   12.44781         832
8   9           Poitier     FRA  46.58329    0.33328       85960
9  10  Clermont-Ferrand     FRA  45.77998    3.08001      233050

Select a subset of columns from the `cities` table and display the first 10 rows.

In [ ]:
%%sql

SELECT name, country FROM cities LIMIT 10;

To select distinct values, use the `DISTINCT` keyword. For example, `SELECT DISTINCT country FROM cities` will return only the distinct values of the `country` column.

In [ ]:
%%sql

SELECT DISTINCT country FROM cities LIMIT 10;

To count the number of rows returned, use the `COUNT(*)` function. For example, `SELECT COUNT(*) FROM cities` will return the number of rows in the `cities` table.

In [18]:
%%sql

SELECT COUNT(*) FROM cities;

count_star()
0          1249

To count the number of distinct values, use the `COUNT(DISTINCT column)` function. For example, `SELECT COUNT(DISTINCT country) FROM cities` will return the number of distinct values in the `country` column.

In [19]:
%%sql

SELECT COUNT(DISTINCT country) FROM cities;

count(DISTINCT country)
0                      200

To calculate the maximum value, use the `MAX(column)` function. For example, `SELECT MAX(population) FROM cities` will return the maximum value in the `population` column.

In [20]:
%%sql

SELECT MAX(population) FROM cities;

max(population)
0         35676000

To calculate the total value, use the `SUM(column)` function. For example, `SELECT SUM(population) FROM cities` will return the total value in the `population` column.

In [21]:
%%sql

SELECT SUM(population) FROM cities;

sum(population)
0     1.475535e+09

To calculate the average value, use the `AVG(column)` function. For example, `SELECT AVG(population) FROM cities` will return the average value in the `population` column.

In [22]:
%%sql

SELECT AVG(population) FROM cities;

avg(population)
0     1.181373e+06

To order the results, use the `ORDER BY column` clause. For example, `SELECT * FROM cities ORDER BY country` will return the rows ordered by the `country` column alphabetically.

In [ ]:
%%sql

SELECT * FROM cities ORDER BY country LIMIT 10;

To order the results in descending order, use the `ORDER BY column DESC` clause. For example, `SELECT * FROM cities ORDER BY country ASC, population DESC` will return the rows ordered by the `country` column alphabetical order and then by the `population` column in descending order.

In [23]:
%%sql

SELECT * FROM cities ORDER BY country ASC, population DESC LIMIT 10;

id            name country  latitude  longitude  population
0  1219           Kabul     AFG  34.51669   69.18326     3277000
1  1024        Kandahar     AFG  31.61002   65.69495      715542
2   665           Herat     AFG  34.33001   62.16999      481009
3   666  Mazar-e Sharif     AFG  36.69999   67.10003      458151
4  1177          Luanda     AGO  -8.83829   13.23443     5172900
5  1027          Huambo     AGO -12.74999   15.76001     1100000
6   670        Benguela     AGO -12.57826   13.40723      151226
7   672          Namibe     AGO -15.19004   12.16002      132900
8   669         Malanje     AGO  -9.54000   16.34003      125856
9   671         Lubango     AGO -14.91001   13.49002      125632

## The WHERE Clause

The `WHERE` clause is used to filter records. The `WHERE` clause is used to extract only those records that fulfill a specified condition.

In [ ]:
%%sql

SELECT * FROM cities WHERE country='USA'

You can use boolean operators such as `AND`, `OR`, `NOT` to filter records. For example, `SELECT * FROM cities WHERE country='USA' OR country='CAN'` will return the rows where the `country` column is either `USA` or `CAN`.

In [24]:
%%sql

SELECT * FROM cities WHERE country='USA' OR country='CAN';

id                 name country  latitude  longitude  population
0     115                Agana     USA  13.47002  144.75002      122411
1     124       San Bernardino     USA  34.12038 -117.30003     1745000
2     125           Bridgeport     USA  41.17998  -73.19996     1018000
3     126            Rochester     USA  43.17043  -77.61995      755000
4     160  International Falls     USA  48.60113  -93.41085       15240
..    ...                  ...     ...       ...        ...         ...
154  1002        Charlottetown     CAN  46.24928  -63.13133       42402
155  1171             Edmonton     CAN  53.55002 -113.49998     1058000
156  1172             Montréal     CAN  45.50000  -73.58330     3678000
157  1216            Vancouver     CAN  49.27342 -123.12164     2313328
158  1217              Toronto     CAN  43.69998  -79.42002     5213000

[159 rows x 6 columns]

To select US cities with a population greater than 1 million, use the following query: `SELECT * FROM cities WHERE country='USA' AND population > 1000000`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country='USA' AND population>1000000;

To select cities with the country name starting with the letter `U`, use the following query: `SELECT * FROM cities WHERE country LIKE 'U%'`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country LIKE 'U%';

To select cities with the country name ending with the letter `A`, use the following query: `SELECT * FROM cities WHERE country LIKE '%A'`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country LIKE '%A';

To select cities with the country name containing the letter `S` in the middle, use the following query: `SELECT * FROM cities WHERE country LIKE '_S_'`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country LIKE '_S_';

To select cities from a list of countries, use the `IN` operator. For example, `SELECT * FROM cities WHERE country IN ('USA', 'CAN')` will return the rows where the `country` column is either `USA` or `CAN`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country IN ('USA', 'CAN');

To select cities with a population between 1 and 10 million, use the following query: `SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000`.

In [ ]:
%%sql

SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000;

## SQL Joins

Reference: https://www.w3schools.com/sql/sql_join.asp

Here are the different types of the JOINs in SQL:

- `(INNER) JOIN`: Returns records that have matching values in both tables
- `LEFT (OUTER) JOIN`: Returns all records from the left table, and the matched records from the right table
- `RIGHT (OUTER) JOIN`: Returns all records from the right table, and the matched records from the left table
- `FULL (OUTER) JOIN`: Returns all records when there is a match in either left or right table

![](https://i.imgur.com/mITYzuS.png)

We have two sample tables: `cities` and `countries`.

There are 1,249 cities in the `cities` table and 243 countries in the `countries` table.

In [25]:
%%sql

SELECT COUNT(*) FROM cities;

count_star()
0          1249

In [26]:
%%sql

SELECT * FROM cities LIMIT 10;

id              name country  latitude  longitude  population
0   1             Bombo     UGA   0.58330   32.53330       75000
1   2       Fort Portal     UGA   0.67100   30.27500       42670
2   3           Potenza     ITA  40.64200   15.79900       69060
3   4        Campobasso     ITA  41.56300   14.65600       50762
4   5             Aosta     ITA  45.73700    7.31500       34062
5   6         Mariehamn     ALD  60.09700   19.94900       10682
6   7          Ramallah     PSE  31.90294   35.20621       24599
7   8      Vatican City     VAT  41.90001   12.44781         832
8   9           Poitier     FRA  46.58329    0.33328       85960
9  10  Clermont-Ferrand     FRA  45.77998    3.08001      233050

In [ ]:
%%sql

SELECT COUNT(*) FROM countries;

In [27]:
%%sql

SELECT * FROM countries LIMIT 10;

id              Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0   1          Afghanistan          AF         AFG             4   33.0000   
1   2              Albania          AL         ALB             8   41.0000   
2   3              Algeria          DZ         DZA            12   28.0000   
3   4       American Samoa          AS         ASM            16  -14.3333   
4   5              Andorra          AD         AND            20   42.5000   
5   6               Angola          AO         AGO            24  -12.5000   
6   7             Anguilla          AI         AIA           660   18.2500   
7   8           Antarctica          AQ         ATA            10  -90.0000   
8   9  Antigua and Barbuda          AG         ATG            28   17.0500   
9  10            Argentina          AR         ARG            32  -34.0000   

   Longitude  
0    65.0000  
1    20.0000  
2     3.0000  
3  -170.0000  
4     1.6000  
5    18.5000  
6   -63.1667  
7     0.0000  
8   -61.8000  
9   -64.0000

### SQL Inner Join

The `INNER JOIN` keyword selects records that have matching values in both tables. In the example, we join the `cities` table with the `countries` table using the `country` column in the `cities` table and the `Alpha3_code` column in the `countries` table. The result contains 1,244 rows, indicating that there are 5 cities that do not have a matching country.

In [31]:
%%sql

SELECT * 
FROM cities 
INNER JOIN countries 
ON cities.country = countries."Alpha3_code";

id         name country  latitude  longitude  population  id_1  \
0        1        Bombo     UGA   0.58330   32.53330       75000   226   
1        2  Fort Portal     UGA   0.67100   30.27500       42670   226   
2        3      Potenza     ITA  40.64200   15.79900       69060   106   
3        4   Campobasso     ITA  41.56300   14.65600       50762   106   
4        5        Aosta     ITA  45.73700    7.31500       34062   106   
...    ...          ...     ...       ...        ...         ...   ...   
1239  1205       Berlin     DEU  52.52182   13.40155     3406000    79   
1240  1208        Osaka     JPN  34.75004  135.46014    11294000   108   
1241  1215       Tehran     IRN  35.67194   51.42434     7873000   101   
1242  1231  Mexico City     MEX  19.44244  -99.13099    19028000   140   
1243  1240        Tokyo     JPN  35.68502  139.75141    35676000   108   

                      Country_1 Alpha2_code Alpha3_code  Numeric_code  \
0                        Uganda          UG         UGA           800   
1                        Uganda          UG         UGA           800   
2                         Italy          IT         ITA           380   
3                         Italy          IT         ITA           380   
4                         Italy          IT         ITA           380   
...                         ...         ...         ...           ...   
1239                    Germany          DE         DEU           276   
1240                      Japan          JP         JPN           392   
1241  Iran, Islamic Republic of          IR         IRN           364   
1242                     Mexico          MX         MEX           484   
1243                      Japan          JP         JPN           392   

      Latitude_1  Longitude_1  
0         1.0000      32.0000  
1         1.0000      32.0000  
2        42.8333      12.8333  
3        42.8333      12.8333  
4        42.8333      12.8333  
...          ...          ...  
1239     51.0000       9.0000  
1240     36.0000     138.0000  
1241     32.0000      53.0000  
1242     23.0000    -102.0000  
1243     36.0000     138.0000  

[1244 rows x 13 columns]

Only select the `city` and `country` columns from the `cities` table and the `Country` column from the `countries` table.

In [30]:
%%sql

SELECT name, cities."country", countries."Country" 
FROM cities 
INNER JOIN countries 
ON cities.country = countries."Alpha3_code";

name country                  Country_1
0           Bombo     UGA                     Uganda
1     Fort Portal     UGA                     Uganda
2         Potenza     ITA                      Italy
3      Campobasso     ITA                      Italy
4           Aosta     ITA                      Italy
...           ...     ...                        ...
1239       Berlin     DEU                    Germany
1240        Osaka     JPN                      Japan
1241       Tehran     IRN  Iran, Islamic Republic of
1242  Mexico City     MEX                     Mexico
1243        Tokyo     JPN                      Japan

[1244 rows x 3 columns]

### SQL Left Join

The `LEFT JOIN` keyword returns all records from the left table (`cities`), and the matched records from the right table (`countries`). The result contains 1,249 rows, the same number of rows as the `cities` table.

In [36]:
%%sql

SELECT *
FROM cities 
LEFT JOIN countries 
ON cities.country = countries."Alpha3_code";

id         name country  latitude  longitude  population   id_1  \
0        1        Bombo     UGA   0.58330   32.53330       75000  226.0   
1        2  Fort Portal     UGA   0.67100   30.27500       42670  226.0   
2        3      Potenza     ITA  40.64200   15.79900       69060  106.0   
3        4   Campobasso     ITA  41.56300   14.65600       50762  106.0   
4        5        Aosta     ITA  45.73700    7.31500       34062  106.0   
...    ...          ...     ...       ...        ...         ...    ...   
1244     6    Mariehamn     ALD  60.09700   19.94900       10682    NaN   
1245   118   Bir Lehlou     SAH  26.11917   -9.65252         500    NaN   
1246   415     Pristina     KOS  42.66671   21.16598      465186    NaN   
1247   719      Berbera     SOL  10.43555   45.01641      242344    NaN   
1248  1053     Hargeysa     SOL   9.56002   44.06531      477876    NaN   

     Country_1 Alpha2_code Alpha3_code  Numeric_code  Latitude_1  Longitude_1  
0       Uganda          UG         UGA         800.0      1.0000      32.0000  
1       Uganda          UG         UGA         800.0      1.0000      32.0000  
2        Italy          IT         ITA         380.0     42.8333      12.8333  
3        Italy          IT         ITA         380.0     42.8333      12.8333  
4        Italy          IT         ITA         380.0     42.8333      12.8333  
...        ...         ...         ...           ...         ...          ...  
1244      None        None        None           NaN         NaN          NaN  
1245      None        None        None           NaN         NaN          NaN  
1246      None        None        None           NaN         NaN          NaN  
1247      None        None        None           NaN         NaN          NaN  
1248      None        None        None           NaN         NaN          NaN  

[1249 rows x 13 columns]

### SQL Right Join

The `RIGHT JOIN` keyword returns all records from the right table (`countries`), and the matched records from the left table (`cities`). The result contains 1,291 rows.

In [37]:
%%sql

SELECT * 
FROM cities 
RIGHT JOIN countries 
ON cities.country = countries."Alpha3_code";

id         name country  latitude  longitude  population  id_1  \
0     1.0        Bombo     UGA    0.5833    32.5333     75000.0   226   
1     2.0  Fort Portal     UGA    0.6710    30.2750     42670.0   226   
2     3.0      Potenza     ITA   40.6420    15.7990     69060.0   106   
3     4.0   Campobasso     ITA   41.5630    14.6560     50762.0   106   
4     5.0        Aosta     ITA   45.7370     7.3150     34062.0   106   
...   ...          ...     ...       ...        ...         ...   ...   
1286  NaN         None    None       NaN        NaN         NaN   185   
1287  NaN         None    None       NaN        NaN         NaN    85   
1288  NaN         None    None       NaN        NaN         NaN   162   
1289  NaN         None    None       NaN        NaN         NaN    83   
1290  NaN         None    None       NaN        NaN         NaN   127   

                      Country_1 Alpha2_code Alpha3_code  Numeric_code  \
0                        Uganda          UG         UGA           800   
1                        Uganda          UG         UGA           800   
2                         Italy          IT         ITA           380   
3                         Italy          IT         ITA           380   
4                         Italy          IT         ITA           380   
...                         ...         ...         ...           ...   
1286  Saint Pierre and Miquelon          PM         SPM           666   
1287                 Guadeloupe          GP         GLP           312   
1288   Northern Mariana Islands          MP         MNP           580   
1289                  Greenland          GL         GRL           304   
1290                      Macao          MO         MAC           446   

      Latitude_1  Longitude_1  
0         1.0000      32.0000  
1         1.0000      32.0000  
2        42.8333      12.8333  
3        42.8333      12.8333  
4        42.8333      12.8333  
...          ...          ...  
1286     46.8333     -56.3333  
1287     16.2500     -61.5833  
1288     15.2000     145.7500  
1289     72.0000     -40.0000  
1290     22.1667     113.5500  

[1291 rows x 13 columns]

### SQL Full Join

The `FULL JOIN` keyword returns all records when there is a match in either left or right table. The result contains 1,296 rows.

In [ ]:
%%sql

SELECT * 
FROM cities 
FULL JOIN countries 
ON cities.country = countries."Alpha3_code";

### SQL Union

The `UNION` operator is used to combine the result-set of two or more `SELECT` statements.

In [ ]:
%%sql

SELECT country FROM cities
UNION
SELECT "Alpha3_code" FROM countries;

## Aggregation

### Group By

The `GROUP BY` statement groups rows that have the same values into summary rows, like "find the number of cities in each country".

The `GROUP BY` statement is often used with aggregate functions (`COUNT`, `MAX`, `MIN`, `SUM`, `AVG`) to group the result-set by one or more columns.

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
ORDER BY COUNT(name) DESC;

In [38]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries 
ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
ORDER BY COUNT(name) DESC;

Country  count("name")
0     United States            114
1             China            100
2            Russia             81
3             India             69
4            Brazil             46
..              ...            ...
192    Sierra Leone              1
193          Jordan              1
194         Jamaica              1
195      Azerbaijan              1
196  Czech Republic              1

[197 rows x 2 columns]

### Having

The `HAVING` clause was added to SQL because the `WHERE` keyword could not be used with aggregate functions.

For example, to select countries with more than 40 cities:

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

In [39]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

Country  count("name")
0  United States            114
1          China            100
2         Russia             81
3          India             69
4         Brazil             46
5         Canada             45

## Conditional statements

The `CASE` statement goes through conditions and returns a value when the first condition is met (like an `IF-THEN-ELSE` statement). So, once a condition is true, it will stop reading and return the result. If no conditions are true, it returns the value in the `ELSE` clause.

For example, to divide cities into 3 groups based on their population:

In [40]:
%%sql

SELECT name, population,
CASE
    WHEN population > 10000000 THEN 'Megacity'
    WHEN population > 1000000 THEN 'Large city'
    ELSE 'Small city'
END AS category
FROM cities;

name  population    category
0              Bombo       75000  Small city
1        Fort Portal       42670  Small city
2            Potenza       69060  Small city
3         Campobasso       50762  Small city
4              Aosta       34062  Small city
...              ...         ...         ...
1244  Rio de Janeiro    11748000    Megacity
1245       Sao Paulo    18845000    Megacity
1246          Sydney     4630000  Large city
1247       Singapore     5183700  Large city
1248       Hong Kong     7206000  Large city

[1249 rows x 3 columns]

## Saving results

You can save the results of a query to a new table using the `CREATE TABLE AS` statement.

In [ ]:
%%sql

CREATE TABLE cities2 AS SELECT * FROM cities;

Show the new table content.

In [ ]:
%%sql

FROM cities2;

Use the `DROP TABLE` statement to delete the table.

In [ ]:
%%sql

DROP TABLE IF EXISTS cities_usa;
CREATE TABLE cities_usa AS (SELECT * FROM cities WHERE country = 'USA');

In [ ]:
%%sql

FROM cities_usa;

Use the `INSERT INTO` statement to insert rows into a table.

In [ ]:
%%sql

INSERT INTO cities_usa (SELECT * FROM cities WHERE country = 'CAN');

## SQL Comments

Comments are used to explain sections of SQL statements, or to prevent execution of SQL statements.


### Single line comMents

Single line comments start with --.

Any text between -- and the end of the line will be ignored (will not be executed).

The following example uses a single-line comment as an explanation:

In [ ]:
%%sql

SELECT * FROM cities LIMIT 10 -- This is a comment;

### Multi-line comments

Multi-line comments start with `/*` and end with `*/`.

Any text between `/*` and `*/` will be ignored.

The following example uses a multi-line comment as an explanation:

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
/*
 * Adding Group by
 * Adding Order by
 */
GROUP BY country
ORDER BY COUNT(name) DESC
LIMIT 10;